pkg/sync/syncer.go

```diff
@@ -1828,6 +1845,26 @@ func (s *syncer) processGrantsWithExternalPrincipals(ctx context.Context, princi
 	expandedGrants := make([]*v2.Grant, 0)
 	for _, grant := range grants {
 		annos := annotations.Annotations(grant.Annotations)
+		matchResourceMatchIDAnno, err := GetExternalResourceMatchIDAnnotation(annos)
+		if err != nil {
+			return err
+		}
+		if matchResourceMatchIDAnno != nil {
+			if principal, ok := principalMap[matchResourceMatchIDAnno.Id]; ok {
+				newGrant := &v2.Grant{
+					Entitlement: grant.Entitlement,
+					Principal:   principal,
+					Id:          batonGrant.NewGrantID(principal, grant.Entitlement),
+					Sources:     grant.Sources,
+					Annotations: grant.Annotations,
+				}
+				expandedGrants = append(expandedGrants, newGrant)
+			}
+			// We still want to delete the grant even if there are no matches
+			grantsToDelete = append(grantsToDelete, grant.Id)
```

Comment:
Why you want to delete the grant that may have matches?? 🫨

Reply:
we want to delete the grant even if there were no matches, these are for grants specifically for principals synced from another connector c1z, so if there are no principals matching our criteria, we delete the grant. 

for example, in gcp-wif if we have a policy that has ```principalSet://iam.googleapis.com/locations/global/workforcePools/pool-id/attribute.email/nonexistent@email.com``` 

and we use the annotation `ExternalResourceMatch{Key: "email", Value: "nonexistent@email.com"}` 

but we have no user with that email, there is no grant we should have here. Similar for `ExternalResourceMatchID`, if we don't have a principal with that ID, we don't have a valid grant.

Reply:
For grants that do have matches, we generate a new grant using the principal that matches.